src/cluster_legacy.c

```diff
@@ -6491,6 +6539,28 @@ void removeChannelsInSlot(unsigned int slot) {
     pubsubShardUnsubscribeAllChannelsInSlot(slot);
 }
 
+/* Propagate deletion of all keys in slot (without doing the local deletion).
+ * Assumed to be executed inside an execution context. */
+unsigned int propagateSlotDeletionByKeys(unsigned int hashslot) {
+    unsigned int j = 0;
+
+    if (!countKeysInSlot(hashslot)) return 0;
+
+    kvstoreHashtableIterator *kvs_di = NULL;
+    void *next;
+    kvs_di = kvstoreGetHashtableIterator(server.db->keys, hashslot, HASHTABLE_ITER_SAFE);
+    while (kvstoreHashtableIteratorNext(kvs_di, &next)) {
+        robj *valkey = next;
+        sds sdskey = objectGetKey(valkey);
+        robj *key = createStringObject(sdskey, sdslen(sdskey));
+        propagateDeletion(&server.db[0], key, server.lazyfree_lazy_server_del);
+        decrRefCount(key);
+        j++;
+    }
+    kvstoreReleaseHashtableIterator(kvs_di);
+    return j;
+}
+
 /* Remove all the keys in the specified hash slot.
  * The number of removed items is returned. */
 unsigned int delKeysInSlot(unsigned int hashslot) {
```

Comment:
A follow up item that I don't want to talk about in the meeting is we should do an async slot purge for this. We don't need to block the main thread for deleting keys. Since we can gate this on the version, we can implement a flushslot to replicate or have the replica also follow.

Reply:
Right. I think we have a separate PR for FLUSHSLOT - although not technically required, it will make it much more usable for real-life workloads where a slot might take quite some time to delete (and propagate deletion to replicas).

Once FLUSHSLOT is in - I think async would make sense - I don't see any major downsides to it.